### ISSUE TITLE
How to map structures from sqlite

### ISSUE THREAD

**Issue by kant2002** [bricelam/SQLitePCL.raw#1]
I think a lot of structures for Sqlitevirtual tables already implemented in blittable way.
- [ ] sqlite3_module
- [x] sqlite3_vtab
- [x] sqlite3_index_info
- [x] sqlite3_vtab_cursor
- [x] sqlite3_index_constraint
- [x] sqlite3_index_orderby
- [x] sqlite3_index_constraint_usage

https://github.com/kant2002/SqliteVirtualTables/blob/40f846a2de62252fa0e9fbb80080af56abafa073/SqliteVirtualTables/VirtualTableSupport.cs#L44-L96

One *problem* with that approach is that it is using unsafe feature which neels a bit unnatural to SqlitePCLRaw, even if I think this is simpler to work with.

So one important question is how we define these auxillary stucture? In unsafe way and closer to C, or in managed way with some Marshal.PtrToStructure magic.

**Comment by bricelam** [comment]
Yeah, I would avoid `unsafe` since that matches the rest of SQLitePCLRaw. Is it the function pointers that are problematic? Here's a "safe" version of `sqlite_vfs` that I wrote a few years ago to get at the xDelete function: https://github.com/aspnet/Microsoft.Data.Sqlite/blob/1.0.0-beta1/src/Microsoft.Data.SQLite/Interop/sqlite3_vfs.cs

**Comment by kant2002** [comment]
`sqlite3_module` by itself probably would be fine, we do it one time and that's it. but some functions there, like 
https://github.com/kant2002/SqliteVirtualTables/blob/e27a2f6fe28d33c039e5c3b8d122d3ff5fa87b6c/Series/SeriesModule.cs#L448-L450

always rely on downcast to impl type, and that probably would be *interesting* in interop. Let's say we define
```csharp
static int seriesRowid(ref sqlite3_vtab_cursor cur, out long pRowid)
{
        ref series_cursor pCur = Unsafe.As<series_cursor>(ref cur);
        ulong n = pCur.ss.uSeqIndexNow;
        pRowid = (long)((n < ulong.MaxValue) ? n + 1 : 0);
        return SQLITE_OK;
}
```

I do not sure that cure is better then disease.

**Comment by bricelam** [comment]
Oh yuck. I haven't used virtual tables much and didn't realize they rely on derived types. Is this the common case or are many implementations possible with just the base struct? If it's common, then yeah, `unsafe` is probably the way to go. Otherwise, we might consider two code paths--one safe and one unsafe.

**Comment by kant2002** [comment]
Yeah, this is common and suggested way of operation.
From: https://www.sqlite.org/c3ref/vtab_cursor.html

> struct sqlite3_vtab_cursor {
>  sqlite3_vtab *pVtab;      /* Virtual table of this cursor */
>  /* Virtual table implementations will typically add additional fields */
>};
> 
> Every [virtual table module](https://www.sqlite.org/c3ref/module.html) implementation uses a subclass of the following structure to describe cursors that point into the [virtual table](https://www.sqlite.org/vtab.html) and are used to loop through the virtual table.

After I wrote answer to you, I do think that maybe "safe" way not so bad, but we have other problem. Sqlite rely on fact that all structures allocated by virtual table module should have static address, so that's mean we should have all objects somehow fixed in memory. I don't know yet how we can do that.

I in general agree that managed way is prefered. Just want to highlight some issues which I discover during my simple prototype.
